storage/src/index/storage.rs

```diff
@@ -0,0 +1,166 @@
+}
+
+impl LocationIterator<'_> {
+    /// Create a `LocationIterator` that returns no items.
+    fn empty() -> Self {
+        LocationIterator { next: None }
+    }
+}
+
+impl Iterator for LocationIterator<'_> {
+    type Item = u64;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.next {
+            Some(next) => {
+                let loc = next.location;
+                self.next = next.next.as_deref();
+                Some(loc)
+            }
+            None => None,
+        }
+    }
+}
+
+impl Record {
+    fn iter(&self) -> LocationIterator {
+        LocationIterator { next: Some(self) }
+    }
+}
+
+pub struct Index<T: Translator, E: Metrics> {
+    translator: T,
+
+    // We store the first location associated with a key within the HashMap entry itself to
+    // significantly reduce the number of random reads we need to do on the heap.
+    map: HashMap<T::Key, Record>,
+
+    collisions: Counter,
+    keys_pruned: Counter,
+
+    _metrics: std::marker::PhantomData<E>,
+}
+
+impl<T: Translator, E: Metrics> Index<T, E> {
+    /// Create a new index.
+    pub fn init(context: E, translator: T) -> Self {
+        let s = Self {
+            translator,
+            map: HashMap::new(),
+            collisions: Counter::default(),
+            keys_pruned: Counter::default(),
+            _metrics: std::marker::PhantomData,
+        };
+        context.register(
+            "keys_pruned",
+            "Number of keys pruned",
+            s.keys_pruned.clone(),
+        );
+
+        s
+    }
+
+    pub fn len(&self) -> usize {
+        self.map.len()
+    }
+
+    pub fn collisions(&self) -> u64 {
+        self.collisions.get()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.map.is_empty()
+    }
+
+    /// Insert a new record into the index.
+    pub fn insert(&mut self, key: &[u8], location: u64) {
+        let translated_key = self.translator.transform(key);
+
+        match self.map.entry(translated_key) {
+            Entry::Occupied(entry) => {
+                let entry: &mut Record = entry.into_mut();
+                entry.next = Some(Box::new(Record {
+                    location,
+                    next: entry.next.take(),
+                }));
+                self.collisions.inc();
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(Record {
+                    location,
+                    next: None,
+                });
+            }
+        };
+    }
+
+    pub fn get(&self, key: &[u8]) -> LocationIterator {
+        let translated_key = self.translator.transform(key);
+        match self.map.get(&translated_key) {
+            Some(head) => head.iter(),
+            None => LocationIterator::empty(),
+        }
+    }
+
+    /// Remove locations associated with the key that match the `prune` predicate.
+    pub fn remove(&mut self, key: &[u8], prune: impl Fn(u64) -> bool) {
+        let translated_key = self.translator.transform(key);
```

Comment:
nit: could add the `// Get translated key` back here

Reply:
This same line is also at the top of get & insert -- should I add the comment to each? (though I find what this line of code is doing to be pretty self-evident FWIW.)